### PR TITLE
:window: Display connection state in connection setting page

### PR DESCRIPTION
**Pull request by evantahler**
This PR displays a Connection's STATE in the UI

pre

### CODE REVIEW

File: airbyte-server/src/main/java/io/airbyte/server/handlers/SchedulerHandler.java
Comment (evantahler) [inline airbyte-server/src/main/java/io/airbyte/server/handlers/SchedulerHandler.java:455]: Rather than do this, I'd rather just call the above `getState()` method, but the input types aren't the same.  Is there a JAVA way to allow for two input types, e.g. `public ConnectionState getState(final ConnectionIdRequestBody | ConnectionUpdateStateBody connectionIdRequestBody)`?

Comment (evantahler) [reply airbyte-server/src/main/java/io/airbyte/server/handlers/SchedulerHandler.java:455]: TODO: Probably needs a test

File: airbyte-webapp/src/components/StateBlock/StateBlock.tsx
Comment (evantahler) [inline airbyte-webapp/src/components/StateBlock/StateBlock.tsx:105]: Any tricks to memoize a method that takes no args?

File: airbyte-webapp/src/components/StateBlock/prismjs.d.ts
Comment (evantahler) [inline airbyte-webapp/src/components/StateBlock/prismjs.d.ts:8]: This is a requirement because of our strict mode

File: airbyte-webapp/src/core/request/AirbyteClient.ts
Comment (evantahler) [inline airbyte-webapp/src/core/request/AirbyteClient.ts:587]: All the changes in this file were auto-generated from the API YML

**Comment by sherifnada** [comment]
This demo looks awesome! FWIW I think the remaining TODOs probably shouldn't block the existing PR IMO (maybe except for validating that the state is a valid JSON rather than a "correct" state object)

**Comment by Phlair** [comment]
This is amazing! I'll defer code review to those more familiar with frontend-y stuff 😅, the really easy UI interface is great though!

Do we want to expose this on cloud / specifically **not** expose this on cloud?
- If we do want to, there'll be some changes needed in cloud's `airbyte-server-wrapped` to add the new API changes here and enable releasing to cloud.
- If not, I'm not sure of the precedent but I know we're hiding other settings like adding custom connectors so I would assume follow similar structure (?)

File: airbyte-server/src/main/java/io/airbyte/server/handlers/SchedulerHandler.java
Comment (pedroslopez) [reply airbyte-server/src/main/java/io/airbyte/server/handlers/SchedulerHandler.java:455]: If not possible, could refactor the common logic to another method that just takes in the UUID, like these:

https://github.com/airbytehq/airbyte/blob/880c759cac06d5eccff2bfb2e51aac8a3c129f70/airbyte-server/src/main/java/io/airbyte/server/handlers/SchedulerHandler.java#L500-L510

**Review by pedroslopez — COMMENTED** [review]
Nice! Just a couple suggestions here, but looking good!

File: airbyte-server/src/main/java/io/airbyte/server/apis/ConfigurationApi.java
Comment (evantahler) [reply airbyte-server/src/main/java/io/airbyte/server/apis/ConfigurationApi.java:752]: ugh.  Java 😓

File: airbyte-server/src/main/java/io/airbyte/server/handlers/SchedulerHandler.java
Comment (sherifnada) [inline airbyte-server/src/main/java/io/airbyte/server/handlers/SchedulerHandler.java:453]: we should add a thing in the docs that says how state can be updated, and makes note of this limitation. Can we return a response that gives the user feedback on the UI about this limitation?

Comment (evantahler) [reply airbyte-server/src/main/java/io/airbyte/server/handlers/SchedulerHandler.java:453]: I'm sure that it's possible, I just need to sort out how :D

**Comment by timroes** [comment]
That looks like an awesome idea. Some thoughts around this I'd like to see before merging this:

## Advanced Mode

Allowing users to edit a raw JSON is always a risky task. I don't think we should have this enabled by default, but force the user to beforehand enable a "advanced debug mode". For that it would be good if we could get under the "General settings" page a new switch to enable this and also giving a description that this might enable parts of the API, that you can break stuff with. (We can iterate on the exact wording on the PR then with Amruta) I think this setting should be persisted per user, since you might not want to have all users with access on a workspace automatically seeing this. I would be fine if we store this flag now in localstorage, if we don't want to add it to the server for now. Also given that we might want this feature in OSS as well as in Cloud, and only cloud currently has a user concept, it might make more sense for now to store it locally. The General Settings page in workspaces is usually "per workspace" which will make this setting a bit weird there. I am also fine if we'd want to create a complete new settings page for it "Advanced Settings" or such.

## Code editor

Since this PR introduces a code-editor I would like to get us to use [Monaco](https://www.npmjs.com/package/@monaco-editor/react) for this. Monaco is the editor used in VSCode as has become a quiet powerful standard editor. While this specific editor might not need a lot of features and would be fine with JSON highlighting, we might need this later on in more places (likes log output), where we want those advanced features. I'd prefer if we then won't need to exchange this place anymore. So if possible, please change this PR to already use Monaco as an editor.

## Where should it show in the UI

I find it a bit weird that it's under the "Settings" tab. Stricly spoken it's not really a setting of the connection imho, but a (temporary/changing) state. Maybe we should add a new tab to this (that's only visible in "Advanced mode") with the name "Debug" that will contain this editor. cc @andyjih your thoughts around the placement?

File: airbyte-webapp/src/components/StateBlock/StateBlock.tsx
Comment (timroes) [reply airbyte-webapp/src/components/StateBlock/StateBlock.tsx:106]: There is no need to memoize this method at all. This method is supposed to be only called once on initialization in the `useEffect` below. It's not passed in to any component that would suddenly rerender unnecessarily by havnig a changed reference here. So this is actually atm only having a performance impact (for the comparison of the useCallback dependencies) on every render, and no benefit.

An additional side-node. the `useMemo` (and therefore `useCallback`) specificaitions are meant to do only performance optimizations, i.e. they don't give a "guarantee" that they won't change despite the dependency array haven't changed. Which - if this would ever change - would mean that we actually call it again in the `useEffect` below and thus actually have a breaking behavior.

File: airbyte-webapp/src/components/StateBlock/StateBlock.tsx
Comment (timroes) [inline airbyte-webapp/src/components/StateBlock/StateBlock.tsx:116]: For all places, if this is a fixed string, please use `id="tables.state.title"` and no template string.

File: airbyte-webapp/src/components/StateBlock/StateBlock.tsx
Comment (timroes) [inline airbyte-webapp/src/components/StateBlock/StateBlock.tsx:21]: We're no longer using styled-component for new code, this should be moved to a SCSS module instead.

File: airbyte-webapp/src/locales/en.json
Comment (timroes) [inline airbyte-webapp/src/locales/en.json:399]: We tend to have i18n ids lower case in the beginning, i.e. `state` not `State`.

File: airbyte-webapp/src/components/StateBlock/StateBlock.tsx
Comment (timroes) [inline airbyte-webapp/src/components/StateBlock/StateBlock.tsx:120]: We should prevent having the editor active before typing, and instead show a loading spinner here while the request still loads. That way we're not running in weird scenarios where the user can already type in the editor while the request is still ongoing. Also Monaco doesn't have a build in error state, thus we might want to solve this separately here. It might be worth already putting this into a separate component in the `components` folder as a new `Editor` component, that could contain the loading state (as well as some other adjustments we'd need to make later to it).

**Review by timroes — DISMISSED** [review]

**Review by edmundito — COMMENTED** [review]
I didn't realize this was a draft so I did a review anyway 😁 - Looking really good, most of my comments were related to aligning with our style but also input safety with forms.

**Comment by evantahler** [comment]
Thanks for the comments @timroes and @edmundito! I've cleaned up most of them, except for the below, which I think are worth talking about:

## Advanced Mode
I don't think that **viewing** your connection's state is "advanced" at all.  The Connectors DX team liked the placement on the settings page, and I think this is where our users expect it as well - see https://github.com/airbytehq/airbyte/issues/3092. I think the placement on the settings page along with the warning modal sufficiently calls out the dangerousness of this form. 

In general I like the notion of a "debug" mode, but I'd want that feature to be fully baked (e.g. persisted in the db, along with some notion of a user account in OSS) before going down that rabbit hole. 
 
## Monaco Editor
I gave the Monaco editor a try, and I don't think we should use it for the following reasons:
1. no way to disable the minimap
2. it attempts to spawn web-workers, which will require us to change our content-policy to support. There are lots of warnings thrown on localhost
3. it's a /heavy/ package with tooltips, formatters, etc... and for this use-case we just don't need that.  

Screen Shot 2022-06-03 at 4 01 38 PM

When we get to log parsing, I wouldn't suggest using a code editor either, because they try to pre-parse the whole file.  As logs can be really long, we are going to need finer-grained control over their parsing and display with some sort of shadow-file - only choosing to format/color/parse certain logs within the viewport (as our current log-renderer does). 

## Where in the UI
Because I don't think we need Advanced mode, I think state *does* belong as a setting - it's where a user configures the connection.

**Comment by andyjih** [comment]
Love it! A couple thoughts:

- Putting this in the Settings tab seems ~fine for the time being and we can iterate in the future. We're going to be revisiting the Connections page design soon so we can revisit then if it makes sense to live somewhere else. My hunch is it belongs somewhere "closer" to the Replication and Streams table as it directly, but that's a discussion for another day.
- What would you think about putting a warning before they can edit the state as well? Too much?

I'll defer to Tim on the Monaco discussion.

**Comment by evantahler** [comment]
> What would you think about putting a warning before they can edit the state as well? Too much?

How would that work in practice?  red text near the form, or perhaps an "unlock" button you have to press first with another modal?

**Comment by evantahler** [comment]
@timroes in the interest of moving this along, I've switched to the Monaco editor.  Thanks for pointing me the right options - looks like reading the typedoc was what I needed to do :D.  The editor looks good, and has some nice hints as you are typing.  There are now quite a few warnings (see screenshot), but I think that getting our CSP in order is a bigger project out of the scope of this PR.  If you are OK merging this with these warnings, OK with me!

Screen Shot 2022-06-06 at 8 39 08 PM

I've made the warnings a little scarier now as well - @andyjih does this work for you?

**Comment by evantahler** [comment]
@timroes I'm also very stumped about the CSS chunk order problem:

```
Failed to compile.
chunk 224 [mini-css-extract-plugin]
Conflicting order. Following module has been added:
 * css ./node_modules/react-scripts/node_modules/css-loader/dist/cjs.js??ruleSet[1].rules[1].oneOf[8].use[1]!./node_modules/react-scripts/node_modules/postcss-loader/dist/cjs.js??ruleSet[1].rules[1].oneOf[8].use[2]!./node_modules/resolve-url-loader/index.js??ruleSet[1].rules[1].oneOf[8].use[3]!./node_modules/react-scripts/node_modules/sass-loader/dist/cjs.js??ruleSet[1].rules[1].oneOf[8].use[4]!./src/components/CodeEditor/CodeEditor.module.scss
despite it was not able to fulfill desired ordering with these modules:
 * css ./node_modules/react-scripts/node_modules/css-loader/dist/cjs.js??ruleSet[1].rules[1].oneOf[6].use[1]!./node_modules/react-scripts/node_modules/postcss-loader/dist/cjs.js??ruleSet[1].rules[1].oneOf[6].use[2]!./node_modules/react-scripts/node_modules/source-map-loader/dist/cjs.js!./src/views/Connector/ConnectorDocumentationLayout/ConnectorDocumentationLayout.module.css
   - couldn't fulfill desired order of chunk group(s)
chunk 224 [mini-css-extract-plugin]
Conflicting order. Following module has been added:
 * css ./node_modules/react-scripts/node_modules/css-loader/dist/cjs.js??ruleSet[1].rules[1].oneOf[8].use[1]!./node_modules/react-scripts/node_modules/postcss-loader/dist/cjs.js??ruleSet[1].rules[1].oneOf[8].use[2]!./node_modules/resolve-url-loader/index.js??ruleSet[1].rules[1].oneOf[8].use[3]!./node_modules/react-scripts/node_modules/sass-loader/dist/cjs.js??ruleSet[1].rules[1].oneOf[8].use[4]!./src/components/StateBlock/StateBlock.module.scss
despite it was not able to fulfill desired ordering with these modules:
 * css ./node_modules/react-scripts/node_modules/css-loader/dist/cjs.js??ruleSet[1].rules[1].oneOf[6].use[1]!./node_modules/react-scripts/node_modules/postcss-loader/dist/cjs.js??ruleSet[1].rules[1].oneOf[6].use[2]!./node_modules/react-scripts/node_modules/source-map-loader/dist/cjs.js!./src/views/Connector/ConnectorDocumentationLayout/ConnectorDocumentationLayout.module.css
   - couldn't fulfill desired order of chunk group(s)
 ```

File: airbyte-api/src/main/openapi/config.yaml
Comment (benmoriceau) [inline airbyte-api/src/main/openapi/config.yaml:1405]: I think we shouldn't have this endpoint. We have a "contract" with the connector which is that the platform never modify the content of a state, it will be only done by the connector. I think that the user will have less context of what is a state than the platform and shouldn't be allowed to modify it. It can lead to a lot of issue like data duplication or data missing.

From where that need come from?

**Review by benmoriceau — DISMISSED** [review]

File: airbyte-api/src/main/openapi/config.yaml
Comment (evantahler) [reply airbyte-api/src/main/openapi/config.yaml:1405]: Looks like this is blowing up more than we initially expected!  I'll add more of the rationale and link back to the user issues up in the description.

Comment (benmoriceau) [reply airbyte-api/src/main/openapi/config.yaml:1405]: I think that we shouldn't make the modification of the state easy.  For technical support and power users, they can update it in the DB directly. I understand the need to display it for clarity purpose but the update has many potential side effect that the user might not fully understand.

**Comment by evantahler** [comment]
This PR is now only about *viewing* the state of your connection.  Due to upcoming changes in how we store state, we will think about editing state in a future issue/PR.

**Comment by evantahler** [comment]
@timroes just noting that the `[mini-css-extract-plugin]` problems are still happening

**Comment by timroes** [comment]
@evantahler I made several adjustments to the FE code:

* Added the required CSP rules
* Removed the left over "editing" code in the `CodeEditor` (we try to not merge unused code, but rather add that code at a later point when it's actually needed)
* Moved the `StateBlock` component from the `components` folder into the place it's used (since this is not a shared basic component, but part of that specific page, it should not be under `components`, which is only meant for shared UI components).
* Turned the query from a `useMutation` into a `useSuspenseQuery`, since it's actually just loading data and not sending it, this is the correct function for it. This also allowed getting rid of all the loading state in the StateBlock component, since it will use React's suspense mechanism for this.
* Removed the description, since it didn't give any additional details over the header, so we won't need it.
* Make the `CodeEditor` component `readOnly` for now (and some minor style adjustments to it).

The only problem persisting is now the mini-css-extract-plugin error, which we'll unfortunately need to solve (in another PR) before we can merge this. I'll start some discussion around this today.

**Comment by timroes** [comment]
> The only problem persisting is now the mini-css-extract-plugin error, which we'll unfortunately need to solve (in another PR) before we can merge this. I'll start some discussion around this today.

I've worked around it since in this PR the only thing we needed the CSS for was the padding in the card, which I moved instead into the card component, so we're not blocking this PR on any further changes to the build system.

**Comment by timroes** [comment]
@evantahler I've passed it on to the rest of the team for review to get a second pair of eyes on it.

**Review by edmundito — COMMENTED** [review]
The code looks good and tested locally to ensure it looks correct.

One concert I spoke with @timroes over slack is that it's a bit confusing that we provide an editor that is never editable, it gives the false sense to a user that the edit functionality needs to be unlocked somewhere.

Since the editing functionality is not available for the time being, I would suggest just switching the current state to a simpler render (e.g.`<pre>` until we have the ability to edit and save the state.

**Comment by evantahler** [comment]
I'm a bit confused about what to do - I think that @timroes wanted to add in the Monaco editor into the codebase for use in future projects and be generally be available for all of our 'looking at code' needs.  This PR was the excuse to do it.  I also feel it's a bit heavy for this component (see the chatter above in this PR).   Should I leave in Monaco or take it out?

**Comment by evantahler** [comment]
After a conversation with @edmundito, here's what we decided:
* For now, just render the state in a `<pre>`, because it's simple and gets the job done
* Leave the Monico editor (and @timroes's CSP changes) in the codebase for use later
* When we come back to this and allow state-editing, we'll switch to the Monico editor (probabily)

**Review by edmundito — APPROVED** [review]

**Pull request merged**
